File: offer30-包含min函数的栈/offer30.go
package main

import (
	"container/list"
)

type MinStack struct {
	stack    *list.List
	minStack *list.List
}

func Constructor() MinStack {
	return MinStack{
		stack:    list.New(),
		minStack: list.New(),
	}
}

func (this *MinStack) Push(x int) {
	this.stack.PushBack(x)
	if this.minStack.Len() == 0 || this.minStack.Back().Value.(int) >= x {
		this.minStack.PushBack(x)
	}
}

func (this *MinStack) Pop() {
	if this.stack.Back().Value.(int) == this.minStack.Back().Value.(int) {
		this.minStack.Remove(this.minStack.Back())
	}
	this.stack.Remove(this.stack.Back())
}

func (this *MinStack) Top() int {
	return this.stack.Back().Value.(int)
}

func (this *MinStack) Min() int {
	return this.minStack.Back().Value.(int)
}

File: offer30-包含min函数的栈/offer30.cpp
#include <iostream>
#include <stack>

using namespace std;

class MinStack {
    stack<int> stack, minStack;
public:
    MinStack() {
        while (!stack.empty()) {
            stack.pop();
        }
        while (!minStack.empty()) {
            minStack.pop();
        }   
    }
    
    void push(int x) {
        stack.push(x);
        if (minStack.empty() || x <= minStack.top())
		    minStack.push(x);
    }
    
    void pop() {
        if (stack.top() == minStack.top())
		    minStack.pop();
        stack.pop();
    }
    
    int top() {
        return stack.top();
    }
    
    int min() {
        return minStack.top();
    }
};
